src/setup.rs

```diff
@@ -207,24 +233,104 @@ pub fn setup(
     )
 }
 
-pub struct AllIOs {
-    pub io: Pin<Gpio10, PushPullOutput>,
-    pub ck: Pin<Gpio11, PushPullOutput>,
-    pub tdi: Pin<Gpio17, PushPullOutput>,
-    pub tdo_swo: Pin<Gpio16, PushPullOutput>,
-    pub reset: Pin<Gpio9, PushPullOutput>,
-    pub vcp_rx: Pin<Gpio21, PushPullOutput>,
-    pub vcp_tx: Pin<Gpio20, PushPullOutput>,
+/// A simple Dynamic pin, meaning it can be output or input.
+pub enum DynPin<P, InputPull>
+where
+    P: PinId,
+    InputPull: PullType,
+{
+    /// The pin is in input mode.
+    Input(Pin<P, FunctionSioInput, InputPull>),
+    /// The pin is in output mode.
+    Output(Pin<P, FunctionSioOutput, InputPull>),
+}
+
+impl<P, InputPull> DynPin<P, InputPull>
+where
+    P: ValidFunction<FunctionSioInput>,
+    P: ValidFunction<FunctionSioOutput>,
+    InputPull: PullType,
+{
+    pub fn into_input(&mut self) {
+        unsafe {
+            replace_with_or_abort_unchecked(self, |p| match p {
+                DynPin::Input(_) => p,
+                DynPin::Output(o) => DynPin::Input(o.into_function().into_pull_type()),
+            })
+        }
+    }
+
+    pub fn into_output(&mut self) {
+        unsafe {
+            replace_with_or_abort_unchecked(self, |p| match p {
+                DynPin::Input(i) => DynPin::Output(i.into_push_pull_output()),
+                DynPin::Output(_) => p,
+            })
+        }
+    }
+
+    pub fn into_output_in_state(&mut self, state: PinState) {
+        unsafe {
+            replace_with_or_abort_unchecked(self, |p| {
+                match p {
+                    DynPin::Input(i) => {
+                        //
+                        DynPin::Output(i.into_push_pull_output_in_state(state))
+                    }
+                    DynPin::Output(_) => p,
+                }
+            })
+        }
+    }
+
+    /// Note: This function panics if the pin is in the wrong mode.
+    pub fn set_high(&mut self) {
+        match self {
+            DynPin::Input(_) => defmt::panic!("Output operation on input pin"),
+            DynPin::Output(o) => {
+                o.set_high().ok();
+            }
+        }
+    }
+
+    /// Note: This function panics if the pin is in the wrong mode.
+    pub fn set_low(&mut self) {
+        match self {
+            DynPin::Input(_) => defmt::panic!("Output operation on input pin"),
+            DynPin::Output(o) => {
+                o.set_high().ok();
```

Comment:
I missed this so many times :facepalm: 
```suggestion
                o.set_low().ok();
```